activesupport/test/cache/stores/mem_cache_store_test.rb

```diff
@@ -34,6 +46,76 @@ def setup
   include EncodedKeyCacheBehavior
   include AutoloadingCacheBehavior
 
+  def test_connection_pool
+    emulating_latency do
+      begin
+        mutex = Mutex.new
+        cache = ActiveSupport::Cache.lookup_store(:mem_cache_store, pool_size: 1, pool_timeout: 1)
+        cache.clear
+
+        thread = Thread.new {
+          mutex.synchronize {
+            cache.read("latency")
+          }
+        }
+
+        # This is an alternative to Thread.pass because, according to Ruby's
+        # documentation, it may or may not pass the execution to another thread.
+        #
+        # Since we can't predict how much time is needed to the thread start
+        # working, we simply wait for its preemption by checking if the mutex
+        # has been acquired.
+        #
+        # This is necessary because we want the read of the previous thread to
+        # happen before the below read.
+        loop do
+          break if mutex.locked?
```

Comment:
I think `Thread.pass until mutex.locked?` is a slightly better spelling here -- same "just keep looping" effect, but it at least *reads* as being less hot.

... though it's worth noting this doesn't guarantee the other thread has started reading before we continue. A `Concurrent::CyclicBarrier` that gets released from *inside* the slow `get` method is probably the safer option.

Reply:
I think we have the same problem with `Concurrent::CyclicBarrier` because you specify the common barrier point but you have no control about who will preempt after that.

The safest option is to use [toxiproxy](https://github.com/Shopify/toxiproxy) to increase the memcached latency at our need and remove it after.

What you think about keeping as is now and implementing toxiproxy in a second PR?

Reply:
Right: it needs to go *inside* the `get` method

Reply:
I got an idea, just a sec.

Reply:
@matthewd take a look now, using `Thread#join` will prevent this problem at all.